reports-scheduler/src/main/java/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/handler/ReportsJobActionHandler.java

```diff
@@ -0,0 +1,192 @@
+    // get all jobs from job_queue index
+    final SearchRequest searchRequest = new SearchRequest(JOB_QUEUE_INDEX_NAME);
+    final SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+    searchSourceBuilder.query(QueryBuilders.matchAllQuery());
+
+    //    // sort the document by enqueue_time
+    //    searchSourceBuilder.sort(new FieldSortBuilder("enqueue_time").order(SortOrder.DESC));
+    searchRequest.source(searchSourceBuilder);
+    client.search(
+        searchRequest,
+        ActionListener.wrap(
+            response -> onSearchJobResponse(response), exception -> onFailure(exception)));
+  }
+
+  private void onSearchJobResponse(SearchResponse response) {
+    final SearchHits hits = response.getHits();
+    final SearchHit[] searchHits = hits.getHits();
+    final List<SearchHit> searchHitsList = Arrays.asList(searchHits);
+    // randomize the jobs, for possible faster job retrieval
+    Randomness.shuffle(searchHitsList);
+    final Queue<SearchHit> searchHitsQueue = new LinkedList<>(searchHitsList);
+
+    log.info("original queue size: " + searchHitsQueue.size());
+
+    findFirstAvailableJob(searchHitsQueue);
+  }
+
+  /**
+   * Attempt to find single jobs that is not being locked, and send the job information back to
+   * reporting core for job execution.
+   *
+   * @param searchHitsQueue a queue which saves all jobs from .reports_scheduler_job_queue index
+   */
+  private void findFirstAvailableJob(Queue<SearchHit> searchHitsQueue) {
+    final SearchHit hit = searchHitsQueue.poll();
+    log.info("queue size after poll: " + searchHitsQueue.size());
+
+    if (hit != null) {
+      String jobId = hit.getId();
+      // set up jobParamater and jobContext that is required by lockService
+      final JobParameter jobParameter = new JobParameter(null, null, null, false, null, null);
+      jobParameter.setLockDurationSeconds(LOCK_DURATION_SECONDS);
+      final JobExecutionContext jobExecutionContext =
+          new JobExecutionContext(null, null, null, JOB_QUEUE_INDEX_NAME, jobId);
+
+      lockService.acquireLock(
+          jobParameter,
+          jobExecutionContext,
+          ActionListener.wrap(
+              lock -> {
+                if (lock == null) {
+                  findFirstAvailableJob(searchHitsQueue);
+                } else {
+                  log.info("send job data(report_definition_id) to reporting core for execution");
+
+                  RestResponse restResponse =
+                      new BytesRestResponse(
+                          RestStatus.OK, hit.toXContent(JsonXContent.contentBuilder(), null));
+                  channel.sendResponse(restResponse);
+                }
+              },
+              exception -> {
+                channel.sendResponse(
+                    new BytesRestResponse(
+                        RestStatus.INTERNAL_SERVER_ERROR, "Failed to acquire lock"));
+                throw new IllegalStateException("Failed to acquire lock.");
```

Comment:
Not sure if throwing exception is the right thing here

Reply:
changed it to `log.debug("Fail to acquire lock " + exception)`